### ISSUE TITLE
Subtyping `.A -> B <= A -> B` leads to wrong `ArgInfo`

### ISSUE THREAD

**Issue by EdNutting** [agda/agda#3986]
Hi,

Replication (using Agda@master and agda-stdlib@master as of today):

```agda
open import Function using (_$_)
open import Data.Nat using (ℕ; _<_; _≤_)
open import Data.Nat.Properties using (≤∧≢⇒<)
open import Data.Fin using (Fin; fromℕ≤; toℕ)
open import Relation.Binary.PropositionalEquality using (_≡_; _≢_)


test-passing₁ : {n m : ℕ} → n ≤ m → n ≢ m → Fin m
test-passing₁ n≤m n≢m = fromℕ≤ (≤∧≢⇒< n≤m n≢m)

test-passing₂ : {n m : ℕ} → n ≤ m → n ≢ m → Fin m
test-passing₂ n≤m n≢m = fromℕ≤ $ ≤∧≢⇒< n≤m n≢m

test-passing₃ : {n m : ℕ} → (n≤m : n ≤ m) → (n≢m : n ≢ m) → n ≡ toℕ (fromℕ≤ (≤∧≢⇒< n≤m n≢m))
test-passing₃ n≤m n≢m
  with {!!}
... | _ = {!!}

test-failing : {n m : ℕ} → (n≤m : n ≤ m) → (n≢m : n ≢ m) → n ≡ toℕ (fromℕ≤ $ ≤∧≢⇒< n≤m n≢m)
test-failing n≤m n≢m
  with ?
... | _ = {!!}
```

Not clear to me what's happening but essentially the test fails when `_$_` is used to provide the irrelevant argument to `fromℕ≤` in the function's type definition (anywhere in it - I just happened to use for the return type here). Also, the test doesn't fail if you remove the `with`.

@mietek helped me with this last night.

Thanks,
Ed

**Comment by andreasabel** [comment]
What exactly is the problem here?  The file loads without complaints, and I see 4 holes:
```
?0 : _38
?1 : n ≡ toℕ (fromℕ≤ (≤∧≢⇒< n≤m n≢m))
?2 : _59
?3 : n ≡ toℕ (fromℕ≤ (≤∧≢⇒< n≤m n≢m))
```
UPDATE: Sorry, wasn't on the latest std-lib.
I can reproduce your error:
```
Expected an irrelevant argument, but found a relevant argument
when checking that the type
{n m : ℕ} (n≤m : n ≤ m) (n≢m : n ≡ m → Data.Empty.⊥) →
_59 (n≤m = n≤m) (n≢m = n≢m) → n ≡ toℕ (fromℕ≤ (≤∧≢⇒< n≤m n≢m))
of the generated with function is well-formed
```

**Comment by gallais** [comment]
Shrunk down:

```agda
open import Agda.Builtin.Equality

postulate
  A : Set
  f : .A → A

_$_ : {A B : Set} → (A → B) → (A → B)
f $ x = f x

test2 : ∀ a → f a ≡ (f $ a)
test2 a with {!!}
... | _ = {!!}
```

**Comment by andreasabel** [comment]
Shrunk down:
```agda
postulate
  A   : Set
  a b : A
  f   : .A → A
  P   : A → Set
  p   : ∀ x → P x

app : (A → A) → A
app g = g a

test-failing : P (app f)
test-failing with b
... | _ = p (f a)
```
The problem is that in `app g`, the application `g a` is relevant, but we are substituting an irrelevant function `f` for `g`, which strictly would need a irrelevant application.
Because `.A -> B` is a subtype of `A -> B`, this is ok, but it means that our relevance annotations in arguments are currently not accurate; they err in the direction of _more relevant_.  This can be bad for the compiler if it relies on the relevance annotation for erasure.

A principled fix seems to require coercive subtyping rather than plain subtyping.  Thus, `f : .A -> B` gets coerced to `\ x -> f .x : A -> B` with the correct relevance annotation in the application `f x`.

**Comment by andreasabel** [comment]
Similar #1186, #2668.  Bug goes back to at least 2.4.2.4.

**Comment by andreasabel** [comment]
This is fixed in 2.6.3 by removal of the `--subtyping` option.
Fixed by:
- https://github.com/agda/agda/pull/5608
